File: packages/golem-ts-sdk/src/internal/mapping/values/unstructured.ts
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Golem Source License v1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { BinaryReference, TextReference } from 'golem:agent/common';

import util from 'node:util';

import { Value } from './Value';

export function serializeBinaryReferenceToValue(tsValue: any): Value {
  const binaryReference = castTsValueToBinaryReference(tsValue);

  switch (binaryReference.tag) {
    case 'url':
      return {
        kind: 'variant',
        caseIdx: 0,
        caseValue: { kind: 'string', value: binaryReference.val },
      };

    case 'inline':
      return {
        kind: 'variant',
        caseIdx: 1,
        caseValue: {
          kind: 'record',
          value: [
            {
              kind: 'list',
              value: Array.from(binaryReference.val.data).map((b) => ({
                kind: 'u8',
                value: b,
              })),
            },
            {
              kind: 'record',
              value: [
                {
                  kind: 'string',
                  value: binaryReference.val.binaryType.mimeType,
                },
              ],
            },
          ],
        },
      };
  }
}

export function serializeTextReferenceToValue(tsValue: any): Value {
  const textReference: TextReference = castTsValueToTextReference(tsValue);

  switch (textReference.tag) {
    case 'url':
      return {
        kind: 'variant',
        caseIdx: 0,
        caseValue: { kind: 'string', value: textReference.val },
      };

    case 'inline':
      if (textReference.val.textType) {
        return {
          kind: 'variant',
          caseIdx: 1,
          caseValue: {
            kind: 'record',
            value: [
              { kind: 'string', value: textReference.val.data },
              {
                kind: 'option',
                value: {
                  kind: 'record',
                  value: [
                    {
                      kind: 'string',
                      value: textReference.val.textType.languageCode,
                    },
                  ],
                },
              },
            ],
          },
        };
      }

      return {
        kind: 'variant',
        caseIdx: 1,
        caseValue: {
          kind: 'record',
          value: [
            { kind: 'string', value: textReference.val.data },
            { kind: 'option' },
          ],
        },
      };
  }
}

export function castTsValueToBinaryReference(tsValue: any): BinaryReference {
  if (typeof tsValue === 'object') {
    const keys = Object.keys(tsValue);

    if (!keys.includes('tag')) {
      throw new Error(
        `Unable to cast value ${util.format(
          tsValue,
        )} to UnstructuredBinary. Missing 'tag' property.`,
      );
    }

    const tag = tsValue['tag'];

    if (typeof tag === 'string' && tag === 'url') {
      if (keys.includes('val')) {
        return {
          tag: 'url',
          val: tsValue['val'],
        };
      } else {
        throw new Error(
          `Unable to cast value ${util.format(
            tsValue,
          )} to UnstructuredBinary. Missing 'val' property for tag 'url'.`,
        );
      }
    }

    if (typeof tag === 'string' && tag === 'inline') {
      if (keys.includes('val') && keys.includes('mimeType')) {
        return {
          tag: 'inline',
          val: {
            data: tsValue['val'],
            binaryType: {
              mimeType: tsValue['mimeType'],
            },
          },
        };
      } else {
        throw new Error(
          `Unable to cast value ${util.format(
            tsValue,
          )} to UnstructuredBinary. Missing 'val' property for tag 'inline'.`,
        );
      }
    }

    throw new Error(
      `Unable to cast value ${util.format(
        tsValue,
      )} to UnstructuredBinary. Invalid 'tag' property: ${tag}. Expected 'url' or 'inline'.`,
    );
  }

  throw new Error(
    `Unable to cast value ${util.format(
      tsValue,
    )} to UnstructuredBinary. Expected an object with 'tag' and 'val' properties.`,
  );
}

export function castTsValueToTextReference(value: any): TextReference {
  if (typeof value === 'object') {
    const keys = Object.keys(value);

    if (!keys.includes('tag')) {
      throw new Error(
        `Unable to cast value ${util.format(
          value,
        )} to UnstructuredText. Missing 'tag' property.`,
      );
    }

    const tag = value['tag'];

    if (typeof tag === 'string' && tag === 'url') {
      if (keys.includes('val')) {
        return {
          tag: 'url',
          val: value['val'],
        };
      } else {
        throw new Error(
          `Unable to cast value ${util.format(
            value,
          )} to UnstructuredText. Missing 'val' property for tag 'url'.`,
        );
      }
    }

    if (typeof tag === 'string' && tag === 'inline') {
      if (keys.includes('val')) {
        if (keys.includes('languageCode')) {
          return {
            tag: 'inline',
            val: {
              data: value['val'],
              textType: {
                languageCode: value['languageCode'],
              },
            },
          };
        } else {
          return {
            tag: 'inline',
            val: {
              data: value['val'],
            },
          };
        }
      } else {
        throw new Error(
          `Unable to cast value ${util.format(
            value,
          )} to UnstructuredText. Missing 'val' property for tag 'inline'.`,
        );
      }
    }

    throw new Error(
      `Unable to cast value ${util.format(
        value,
      )} to UnstructuredText. Invalid 'tag' property: ${tag}. Expected 'url' or 'inline'.`,
    );
  }

  throw new Error(
    `Unable to cast value ${util.format(
      value,
    )} to UnstructuredText. Expected an object with 'tag' and 'val' properties.`,
  );
}

File: cli/golem-cli/src/bridge_gen/scala/runtime/golem/bridge/runtime/Unstructured.scala
/*
 * Copyright 2024-2026 Golem Cloud
 *
 * Licensed under the Golem Source License v1.1 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://license.golem.cloud/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golem.bridge.runtime

import golem.bridge.runtime.SchemaValue._

/**
 * Ergonomic wrapper for the role-marked unstructured-text schema type
 * (`variant { inline: text<restrictions>, url: url }`), mirroring the Golem
 * TypeScript bridge's `UnstructuredText`. A value is either inline text (with an
 * optional language code) or a URL reference.
 */
sealed trait UnstructuredText extends Product with Serializable
object UnstructuredText {
  final case class Inline(value: String, languageCode: Option[String]) extends UnstructuredText
  final case class Url(value: String)                                  extends UnstructuredText

  /** Inline unstructured text with an optional BCP-47 language code. */
  def fromInline(value: String, languageCode: Option[String] = None): UnstructuredText =
    Inline(value, languageCode)

  /** A URL reference to unstructured text. */
  def fromUrl(url: String): UnstructuredText = Url(url)

  // Variant case indices of the canonical role-marked unstructured wrapper:
  // `variant { inline: text, url: url }`.
  private val InlineCase = 0
  private val UrlCase    = 1

  /** Encodes into the schema-native `variant { inline, url }` value. */
  def toSchemaValue(input: UnstructuredText): SchemaValue = input match {
    case Inline(value, languageCode) =>
      VariantValue(InlineCase, Some(TextValue(value, languageCode)))
    case Url(url) =>
      VariantValue(UrlCase, Some(UrlValue(url)))
  }

  /**
   * Decodes a schema-native unstructured-text `variant { inline, url }` value,
   * validating the language tag against `allowedCodes` when the agent declares
   * a fixed set. A missing language is always allowed (lenient decode, matching
   * the server's `check_text`).
   */
  def fromSchemaValue(
    parameterName: String,
    value: SchemaValue,
    allowedCodes: List[String]
  ): Either[String, UnstructuredText] = value match {
    case VariantValue(UrlCase, Some(UrlValue(url))) =>
      Right(Url(url))
    case VariantValue(InlineCase, Some(TextValue(text, language))) =>
      if (allowedCodes.nonEmpty && language.exists(l => !allowedCodes.contains(l)))
        Left(
          s"Invalid value for parameter $parameterName. Language code `${language.get}` is not allowed. " +
            s"Allowed codes: ${allowedCodes.mkString(", ")}"
        )
      else
        Right(Inline(text, language))
    case other =>
      Left(s"Invalid value for parameter $parameterName. Expected an unstructured-text variant, got $other")
  }
}

/**
 * Ergonomic wrapper for the role-marked unstructured-binary schema type
 * (`variant { inline: binary<restrictions>, url: url }`), mirroring the Golem
 * TypeScript bridge's `UnstructuredBinary`. A value is either inline bytes (with
 * an optional MIME type) or a URL reference.
 */
sealed trait UnstructuredBinary extends Product with Serializable
object UnstructuredBinary {
  final case class Inline(bytes: Vector[Byte], mimeType: Option[String]) extends UnstructuredBinary
  final case class Url(value: String)                                    extends UnstructuredBinary

  /** Inline unstructured bytes with an optional MIME type. */
  def fromInline(bytes: Vector[Byte], mimeType: Option[String] = None): UnstructuredBinary =
    Inline(bytes, mimeType)

  /** A URL reference to unstructured binary content. */
  def fromUrl(url: String): UnstructuredBinary = Url(url)

  private val InlineCase = 0
  private val UrlCase    = 1

  /** Encodes into the schema-native `variant { inline, url }` value. */
  def toSchemaValue(input: UnstructuredBinary): SchemaValue = input match {
    case Inline(bytes, mimeType) =>
      VariantValue(InlineCase, Some(BinaryValue(bytes, mimeType)))
    case Url(url) =>
      VariantValue(UrlCase, Some(UrlValue(url)))
  }

  /** Decodes a schema-native unstructured-binary `variant { inline, url }` value. */
  def fromSchemaValue(
    parameterName: String,
    value: SchemaValue
  ): Either[String, UnstructuredBinary] = value match {
    case VariantValue(UrlCase, Some(UrlValue(url))) =>
      Right(Url(url))
    case VariantValue(InlineCase, Some(BinaryValue(bytes, mimeType))) =>
      Right(Inline(bytes, mimeType))
    case other =>
      Left(s"Invalid value for parameter $parameterName. Expected an unstructured-binary variant, got $other")
  }
}
